ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -1877,6 +1938,19 @@ castEarlyExitIntersects =
         Nothing                 -> NoLongerIntersects
         Just (Intersects kis a) -> StillIntersects a kis
 
+-- | Whether we just negotiated an intersection. This is used to specially
+-- handle the first @MsgRollBackward@ that is sent right after.
+data WhetherJustFoundIntersection blk =
+    -- | We just received and successfully processed a @MsgIntersectFound@. The
+    -- ChainSync network spec guarantees that the reply to the next
+    -- @MsgRequestNext@ will be @MsgRollBackward@ to the intersection point.
+    JustFoundIntersection !(Point blk)
+  | -- | The last intersection was negotiated in the past, ie we received at
+    -- least one reply to @MsgRequestNext@ in the meantime.
+    IntersectionWasFoundInThePast
```

Comment:
I suggest renaming this, since "the past" is a relatively ambiguous term, easy to interpret in a variety of ways

Perhaps: `ExpectingRequiredMsgRollBack | NotExpectingRequiredMsgRollBack`

Reply:
Always grateful for better naming suggestions, will change. (I dropped the `Required` for now as this PR does not yet actually *require* the rollback, it just tolerates it.)

Reply:
Superseded by using the approach from https://github.com/IntersectMBO/ouroboros-consensus/pull/1186#discussion_r1688176610